torchtune/modules/loss/ce_chunked_output_loss.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List
+
+import torch
+import torch.nn.functional as F
+
+
+class CEWithChunkedOutputLoss(torch.nn.Module):
+    """
+    CE with chunked outputs that saves memory by only upcasting one chunk at a time.
+
+    Since the model is trained with fp16, before running CE, we have to upcast
+    it to fp32 for better accuracy and stability. When upcasting happens, the memory usage doubles.
+    Models like llama3 have large vocabulary size and, therefore, have a large output
+    result (bsz, num_tokens, vocab_size). If we chunk on the token level, you can still compute
+    the cross entropy normally, but upcasting only one chunk at a time saves considerable memory.
+
+    The CE and upcasting have to be compiled together for better performance.
+    Also, compiling CE always yields great memory gains.
```

Comment:
nit: I would leave absolutes like this out of the docstring (fine to say something like "When using this class, we recommend compiling either `_compute_cross_entropy` or the entire class for reduced memory")